packages/netlify-cms-backend-azure/src/implementation.ts

```diff
@@ -0,0 +1,398 @@
+    initialWorkflowStatus: string;
+  };
+  identityUrl: string;
+  repo: AzureRepo;
+  branch: string;
+  apiRoot: string;
+  token: string | null;
+  squashMerges: boolean;
+  mediaFolder: string;
+  previewContext: string;
+
+  _mediaDisplayURLSem?: Semaphore;
+
+  constructor(config: Config, options = {}) {
+    this.options = {
+      proxied: false,
+      API: null,
+      initialWorkflowStatus: '',
+      ...options,
+    };
+
+    if (!this.options.proxied) {
+      if (config.backend.repo === null || config.backend.repo === undefined) {
+        throw new Error('The Azure backend needs a "repo" in the backend configuration.');
+      }
+    }
+
+    this.api = this.options.API || null;
+
+    this.repo = new AzureRepo(config.backend.repo);
+    this.branch = config.backend.branch || 'master';
+    this.identityUrl = config.backend.identity_url || '';
+    this.apiRoot = config.backend.api_root || 'https://dev.azure.com';
+    this.token = '';
+    this.squashMerges = config.backend.squash_merges || false;
+    this.mediaFolder = trim(config.media_folder, '/');
+    this.previewContext = config.backend.preview_context || '';
+    this.lock = asyncLock();
+  }
+
+  authComponent() {
+    return AuthenticationPage;
+  }
+
+  restoreUser(user: User) {
+    return this.authenticate(user);
+  }
+
+  /**
+   * Handle authentication by creating the Azure DevOps API object, which
+   * will be used for future operations. Also load details of the current
+   * user.
+   *
+   * @param state Oauth credentials containing implicit flow auth token.
+   */
+  async authenticate(state: Credentials) {
+    this.token = state.token as string;
+    this.api = new API(
+      {
+        apiRoot: this.apiRoot,
+        repo: this.repo,
+        branch: this.branch,
+        path: '/',
+        squashMerges: this.squashMerges,
+        initialWorkflowStatus: this.options.initialWorkflowStatus,
+      },
+      this.token,
+    );
+
+    // Get a JSON representation of the user object and capture the name and email
+    // address for later use on commits.
+    const user = await this.api.user();
+    this.api!.commitAuthor = { name: user.displayName, email: user.emailAddress };
+    return { name: user.displayName, token: state.token as string, ...user };
+  }
+
+  /**
+   * Log the user out by forgetting their access token.
+   * TODO: *Actual* logout by redirecting to:
+   * https://login.microsoftonline.com/{tenantId}/oauth2/logout?client_id={clientId}&post_logout_redirect_uri={baseUrl}
+   */
+  logout() {
+    this.token = null;
+    return;
+  }
+
+  getToken() {
+    return Promise.resolve(this.token);
+  }
+
+  entriesByFolder(collection: string, extension: string, depth: number) {
+    return this.api!.listFiles(collection)
+      .then(files => {
+        if (extension) {
+          return files.filter(file => file.relativePath.endsWith('.' + extension));
+        }
+        return files;
+      })
+      .then(this.fetchFiles);
```

Comment:
See https://github.com/netlify/netlify-cms/blob/684b79e43bebb63ce1e844eae5c8c0e76087687b/packages/netlify-cms-backend-bitbucket/src/implementation.ts#L261 for using the utility method `entriesByFolder` to avoid duplication of `fetchFiles` (https://github.com/netlify/netlify-cms/blob/684b79e43bebb63ce1e844eae5c8c0e76087687b/packages/netlify-cms-lib-util/src/implementation.ts#L166)